narrow-filters: Add support for "channel" and "channels" operators.

laurynmm opened the pull request:
Adds backend support for `channel` and `channels` as search/narrow operators and updates the web-app to start using these operators in the `Filter` class, without making user-facing changes to search suggestion strings and URLs.

In the backend changes:
- There's a prep commit to make the test changes easier to read.
- Implements each operator in a separate commit.
- Since the backend/server code isn't user-facing, I went ahead and updated `zerver/lib/narrow.py` and `zerver/tests/test_message_fetch.py` for the stream to channel rename.

In the frontend changes:
- There's a prep commit to make the test changes easier to read.
- Implements each operator in a separate commit.
- The search suggestions still use the "stream" and "stream" strings and neither of the new operators will offer suggestions until a colon is added, e.g. `channel:` is input into the search field.
- URLs still all use "stream" and "streams" with these changes even though the `Filter` class has been updated.

The final commit is the API documentation update and version bump. There are no help center or general API documentation updates in these changes, as those will be done when the user-facing strings are updated.

Part of the stream to channel rename project. See [relevant CZO](https://chat.zulip.org/#narrow/stream/438-release-management/topic/rename.20.22stream.22.20to.20.22channel.22.20project.2C.20.2328468/near/1758864) discussion.

---

**Screenshots and screen captures:**

<details>
<summary>API changelog entry</summary>

![Screenshot from 2024-04-09 20-40-09](https://github.com/zulip/zulip/assets/63245456/7498426b-0d6c-438e-b545-6f8463fc97e1)
</details>
<details>
<summary>Construct a narrow updated "Changes" note</summary>

[Current documentation](https://zulip.com/api/construct-narrow)
![Screenshot from 2024-04-09 20-41-36](https://github.com/zulip/zulip/assets/63245456/a3388ca2-380c-48b7-accb-8d1d9a3771aa)
</details>
<details>
<summary>Get messages endpoint - narrow parameter</summary>

[Current documentation](https://zulip.com/api/get-messages#parameter-narrow)
![Screenshot from 2024-04-09 20-40-28](https://github.com/zulip/zulip/assets/63245456/baffc12c-ef48-4759-8555-3e71ae3f0779)
</details>
<details>
<summary>Update message flags endpoint - narrow parameter</summary>

[Current documentation](https://zulip.com/api/update-message-flags-for-narrow#parameter-narrow)
![Screenshot from 2024-04-09 20-40-56](https://github.com/zulip/zulip/assets/63245456/bdd4305b-c603-4214-b7d9-afdfd876e0f2)
</details>
<details>
<summary>Check messages match narrow endpoint - narrow parameter</summary>

[Current documentation](https://zulip.com/api/check-messages-match-narrow#parameter-narrow)
![Screenshot from 2024-04-09 20-41-08](https://github.com/zulip/zulip/assets/63245456/00d985be-ff8b-4458-91e9-2b110ce589db)
</details>
<details>
<summary>Register queue endpoint - narrow parameter</summary>

[Current documentation](https://zulip.com/api/register-queue#parameter-narrow)
![Screenshot from 2024-04-09 20-41-18](https://github.com/zulip/zulip/assets/63245456/33d2cb5f-f2e5-46e5-b1e9-d2cd35cee7c8)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on web/src/util.ts:
```diff
@@ -281,10 +281,17 @@ export function is_stream_synonym(text: string): boolean {
+    return text === "channels";
+}
+
+export function canonicalize_stream_synonyms(text: string): string {
```
When this is called either only one of the two possible strings will ever be passed to it or either one will and therefore the correct singular/plural of "stream" needs to be returned.

timabbott replied:
This function probably deserves a comment noting its purpose -- for parts of the codebase that have been converted to use `channel/channels` internally, this is used to convert those back into `stream/streams` for external presentation.

laurynmm replied:
Added this comment to the function in the updated changes.

laurynmm commented on web/tests/hashchange.test.js:
```diff
@@ -78,6 +79,29 @@ run_test("terms_round_trip", () => {
     assert.deepEqual(narrow, [{operator: "stream", operand: "Florida, USA", negated: false}]);
 });
 
+run_test("stream_to_channel_rename", () => {
```
Added this test to make it clear in this interim state what's happening with these functions as they interact together.

timabbott replied:
Yeah, this is very helpful.

laurynmm commented:
CI failed due to a [known issue](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/main.20failing/near/1777608) unrelated to these changes.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -408,47 +425,49 @@ def _pg_re_escape(self, pattern: str) -> str:
+            # private channels you are no longer subscribed to, we
             # need get_stream_by_narrow_operand_access_unchecked here.
-            stream = get_stream_by_narrow_operand_access_unchecked(operand, self.realm)
+            channel = get_stream_by_narrow_operand_access_unchecked(operand, self.realm)
```
This local variable rename may be a bit risky if we're not also renaming `class Stream`.

laurynmm replied:
I did these backend commits early on in the rename process and was in the focused on deprecating the stream and streams operators. So, the only remaining cases of `stream` in this file are either functions from  other files, references to the legacy operators or the message type.

<details>
<summary>git grep of "stream" in narrow.py after these changes</summary>

```console
$ git grep stream zerver/lib/narrow.py
zerver/lib/narrow.py:from zerver.lib.streams import (
zerver/lib/narrow.py:    can_access_stream_history_by_id,
zerver/lib/narrow.py:    can_access_stream_history_by_name,
zerver/lib/narrow.py:    get_public_streams_queryset,
zerver/lib/narrow.py:    get_stream_by_narrow_operand_access_unchecked,
zerver/lib/narrow.py:    get_web_public_streams_queryset,
zerver/lib/narrow.py:from zerver.models.streams import get_active_streams
zerver/lib/narrow.py:# "stream" is a legacy alias for "channel"
zerver/lib/narrow.py:channel_operators: List[str] = ["channel", "stream"]
zerver/lib/narrow.py:# "streams" is a legacy alias for "channels"
zerver/lib/narrow.py:channels_operators: List[str] = ["channels", "streams"]
zerver/lib/narrow.py:        # or term == {'operator': 'streams', 'operand': 'web-public', 'negated': False}
zerver/lib/narrow.py:                if message["type"] != "stream":
zerver/lib/narrow.py:                if message["type"] != "stream":
zerver/lib/narrow.py:                if message["type"] != "stream":
zerver/lib/narrow.py:            # "stream" is a legacy alias for "channel"
zerver/lib/narrow.py:            "stream": self.by_channel,
zerver/lib/narrow.py:            # "streams" is a legacy alias for "channels"
zerver/lib/narrow.py:            "streams": self.by_channels,
zerver/lib/narrow.py:                | UserMessage.flags.stream_wildcard_mentioned.mask
zerver/lib/narrow.py:            # need get_stream_by_narrow_operand_access_unchecked here.
zerver/lib/narrow.py:            channel = get_stream_by_narrow_operand_access_unchecked(operand, self.realm)
zerver/lib/narrow.py:            matching_channels = get_active_streams(self.realm).filter(
zerver/lib/narrow.py:            recipient_queryset = get_public_streams_queryset(self.realm)
zerver/lib/narrow.py:            recipient_queryset = get_web_public_streams_queryset(self.realm)
zerver/lib/narrow.py:                    include_history = can_access_stream_history_by_name(user_profile, operand)
zerver/lib/narrow.py:                    include_history = can_access_stream_history_by_id(user_profile, operand)
zerver/lib/narrow.py:                and user_profile.can_access_public_streams()
zerver/lib/narrow.py:                return get_stream_by_narrow_operand_access_unchecked(term["operand"], realm)
```
</details>

laurynmm replied:
Made no changes here in the most recent updates.

timabbott replied:
Yeah this seems fine to think of as a "file-level" transition.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -97,20 +97,35 @@ def read_stop_words() -> List[str]:
 
 
+# "stream" is a legacy alias for "channel"
+channel_operators: List[str] = ["channel", "stream"]
```
Do we need to add something that banes passing both `channel` and `stream` in the same narrow construction with different values?

laurynmm replied:
So, at the moment if we fetch messages with a narrow with two different streams, it doesn't return an error. It just returns no matching messages. After these changes, it's the same for a narrow with two channel operators or a channel and a stream operator (with different operands).

Therefore, if want to ban doing that, we should probably do so as a prep commit and that might also possibly merit a changelog entry.

timabbott replied:
OK, seems fine to take no action on this for now.

timabbott commented on web/src/filter.ts:
```diff
@@ -442,7 +454,8 @@ export class Filter {
             }
-            return sign + term.operator + ":" + Filter.encodeOperand(term.operand.toString());
+            const operator = util.canonicalize_stream_synonym(term.operator);
+            return sign + operator + ":" + Filter.encodeOperand(term.operand.toString());
```
Would it be correct to always canonicalize operators here? I wonder if just calling `canonicalize_operator` here could be a good prep commit.

timabbott replied:
The answer is no -- because we want to still generate `streams:` for the search box, OK.

timabbott commented on web/src/filter.ts:
```diff
@@ -562,9 +575,9 @@ export class Filter {
         if (terms.length >= 2) {
             const is = (term: NarrowTerm, expected: string): boolean =>
-                term.operator === expected && !term.negated;
+                util.canonicalize_stream_synonym(term.operator) === expected && !term.negated;
```
This could probably safely be `self.canonicalize_operator` as a nicer abstraction.

laurynmm replied:
Updated to use `Filter.canonicalize_operator` here.

timabbott commented on web/src/filter.ts:
```diff
@@ -1100,18 +1118,19 @@ export class Filter {
     }
 
     filter_with_new_params(params: NarrowTerm): Filter {
+        const new_params = this.fix_terms([params])[0];
```
This change could probably be a prep commit, since it deserves its own commit message about how we need to use canonical version of terms to do the adjustments correctly.

laurynmm replied:
Separated this out into a separate prep commit.

timabbott commented on web/src/hash_util.ts:
```diff
@@ -93,7 +94,7 @@ export function search_terms_to_hash(terms?: NarrowTerm[]): string {
         for (const term of terms) {
             // Support legacy tuples.
-            const operator = term.operator;
+            const operator = util.canonicalize_stream_synonym(term.operator);
```
This may not be required, in that I believe all callers are the output of `filter.terms()`, which has already canonicalized this? Oh, or maybe it is because we're using `stream` in the URLs and `channel` in the internals, in the temporary state? OK.

timabbott commented:
This looks good though I've not tried to audit for whether there's additional places that need similar logic; I can do some auditing for that myself, but am also expecting that we'll 

Kinda crazy that this reduces references to `git grep stream` by only 0.7% (total of 43k matches); just gives a sense of what the scope might be for renaming the model in the code itself. I think we'll have to figure out something tool-assisted for that sort of work, if we decide to invest in it.

laurynmm commented:
**Updates**:
- Updated for review feedback, see individual comments above.
- Updated the API changelog feature level to 249. The screenshots above are all correct except for the feature level number.
- Fixed up the commit messages a bit.

---

> This looks good though I've not tried to audit for whether there's additional places that need similar logic; I can do some auditing for that myself, but am also expecting that we'll

I'm not sure what you're final thought was here. What other places were you thinking of to audit?
 
> Kinda crazy that this reduces references to `git grep stream` by only 0.7% (total of 43k matches); just gives a sense of what the scope might be for renaming the model in the code itself. I think we'll have to figure out something tool-assisted for that sort of work, if we decide to invest in it.

Yeah, I agree with that assessment. It's taken a lot more focus to make this transition than what I recall from the "private" to "direct" rename.

timabbott commented:
This looks great, rebased to renumber 249 => 250 and marked to merge once CI passes, thanks @laurynmm!

I'll try to get this on chat.zulip.org today so that we can start discovering any quirks in the migrations.

mateuszmandera commented:
@timabbott Shouldn't this also bump `API_FEATURE_LEVEL` to 250? It's still at 249

laurynmm commented:
Opened #29735 to bump the API version.
